File: example-jv-04retrofit/src/main/java/foo/bar/example/foreretrofit/feature/fruit/FruitFetcher.java
package foo.bar.example.foreretrofit.feature.fruit;

import java.util.List;
import java.util.Random;

import co.early.fore.core.Affirm;
import co.early.fore.core.WorkMode;
import co.early.fore.core.callbacks.FailureCallbackWithPayload;
import co.early.fore.core.callbacks.SuccessCallback;
import co.early.fore.core.logging.Logger;
import co.early.fore.core.observer.ObservableImp;
import co.early.fore.net.retrofit2.CallProcessorRetrofit2;
import foo.bar.example.foreretrofit.api.fruits.FruitPojo;
import foo.bar.example.foreretrofit.api.fruits.FruitService;
import foo.bar.example.foreretrofit.api.fruits.FruitsCustomError;
import foo.bar.example.foreretrofit.message.UserMessage;

/**
 * gets a list of fruit from the network, selects one at random to be currentFruit
 */
public class FruitFetcher extends ObservableImp{

    public static final String TAG = FruitFetcher.class.getSimpleName();

    private final FruitService fruitService;
    private final CallProcessorRetrofit2<UserMessage> callProcessor;
    private final WorkMode workMode;
    private final Logger logger;
    private static Random random = new Random();

    private boolean busy;
    private FruitPojo currentFruit = new FruitPojo("(fruitless)", false, 0);

    public FruitFetcher(FruitService fruitService, CallProcessorRetrofit2<UserMessage> callProcessor, Logger logger, WorkMode workMode) {
        super(workMode);
        this.fruitService = Affirm.notNull(fruitService);
        this.callProcessor = Affirm.notNull(callProcessor);
        this.logger = Affirm.notNull(logger);
        this.workMode = Affirm.notNull(workMode);
    }


    public void fetchFruits(final SuccessCallback successCallback, final FailureCallbackWithPayload<UserMessage> failureCallbackWithPayload){

        logger.i(TAG, "fetchFruits()");

        Affirm.notNull(successCallback);
        Affirm.notNull(failureCallbackWithPayload);

        if (busy){
            failureCallbackWithPayload.fail(UserMessage.ERROR_BUSY);
            return;
        }

        busy = true;
        notifyObservers();

        callProcessor.processCall(fruitService.getFruitsSimulateOk("3s"), workMode, FruitsCustomError.class,
                successResponse -> handleSuccess(successCallback, successResponse),
                failureMessage -> handleFailure(failureCallbackWithPayload, failureMessage));

    }


    //identical to fetchFruits() but for demo purposes the URL we point to will give us an error, we also don't specify a custom error class here
    public void fetchFruitsButFailBasic(final SuccessCallback successCallback, final FailureCallbackWithPayload<UserMessage> failureCallbackWithPayload){

        logger.i(TAG, "fetchFruitsButFailBasic()");

        Affirm.notNull(successCallback);
        Affirm.notNull(failureCallbackWithPayload);

        if (busy){
            failureCallbackWithPayload.fail(UserMessage.ERROR_BUSY);
            return;
        }

        busy = true;
        notifyObservers();

        callProcessor.processCall(fruitService.getFruitsSimulateNotAuthorised("3s"), workMode,
                successResponse -> handleSuccess(successCallback, successResponse),
                failureMessage -> handleFailure(failureCallbackWithPayload, failureMessage));

    }


    //identical to fetchFruits() but for demo purposes the URL we point to will give us an error
    public void fetchFruitsButFailAdvanced(final SuccessCallback successCallback, final FailureCallbackWithPayload<UserMessage> failureCallbackWithPayload){

        logger.i(TAG, "fetchFruitsButFailAdvanced()");

        Affirm.notNull(successCallback);
        Affirm.notNull(failureCallbackWithPayload);

        if (busy){
            failureCallbackWithPayload.fail(UserMessage.ERROR_BUSY);
            return;
        }

        busy = true;
        notifyObservers();

        callProcessor.processCall(fruitService.getFruitsSimulateNotAuthorised("3s"), workMode, FruitsCustomError.class,
                successResponse -> handleSuccess(successCallback, successResponse),
                failureMessage -> handleFailure(failureCallbackWithPayload, failureMessage));

    }

    private void handleSuccess(SuccessCallback successCallBack, List<FruitPojo> successResponse){
        currentFruit = selectRandomFruit(successResponse);
        successCallBack.success();
        complete();
    }

    private void handleFailure(FailureCallbackWithPayload<UserMessage> failureCallbackWithPayload, UserMessage failureMessage){
        failureCallbackWithPayload.fail(failureMessage);
        complete();
    }

    public boolean isBusy() {
        return busy;
    }

    public FruitPojo getCurrentFruit() {
        return currentFruit;
    }

    private void complete(){

        logger.i(TAG, "complete()");

        busy = false;
        notifyObservers();
    }

    private FruitPojo selectRandomFruit(List<FruitPojo> listOfFruits){
        return listOfFruits.get(listOfFruits.size() == 1 ? 0 : random.nextInt(listOfFruits.size()-1));
    }

}

File: app-examples/example-kt-04retrofit/src/main/java/foo/bar/example/foreretrofitkt/feature/fruit/FruitFetcher.kt
package foo.bar.example.foreretrofitkt.feature.fruit

import co.early.fore.core.observer.Observable
import co.early.fore.kt.core.coroutine.awaitMain
import co.early.fore.kt.core.coroutine.launchMain
import co.early.fore.kt.core.logging.Logger
import co.early.fore.kt.core.observer.ObservableImp
import co.early.fore.kt.core.type.Either.Fail
import co.early.fore.kt.core.type.Either.Success
import co.early.fore.kt.net.retrofit2.CallWrapperRetrofit2
import co.early.fore.kt.net.retrofit2.carryOn
import foo.bar.example.foreretrofitkt.api.fruits.FruitPojo
import foo.bar.example.foreretrofitkt.api.fruits.FruitService
import foo.bar.example.foreretrofitkt.api.fruits.FruitsCustomError
import foo.bar.example.foreretrofitkt.message.ErrorMessage
import java.util.*

typealias SuccessCallback = () -> Unit
typealias FailureCallback<T> = (T) -> Unit

/**
 * gets a list of fruit from the network, selects one at random to be currentFruit
 */
class FruitFetcher(
    private val fruitService: FruitService,
    private val callWrapper: CallWrapperRetrofit2<ErrorMessage>,
    private val logger: Logger
) : Observable by ObservableImp(logger = logger) {

    var isBusy: Boolean = false
        private set
    var currentFruit = FruitPojo("(fruitless)", false, 0)
        private set


    fun fetchFruitsAsync(
        success: SuccessCallback,
        failureWithPayload: FailureCallback<ErrorMessage>
    ) {

        logger.i("fetchFruitsAsync() t:" + Thread.currentThread())

        if (isBusy) {
            failureWithPayload(ErrorMessage.ERROR_BUSY)
            return
        }

        isBusy = true
        notifyObservers()

        launchMain {

            logger.i("about to use CallProcessor t:" + Thread.currentThread())

            val deferredResult = callWrapper.processCallAsync {

                logger.i("processing call t:" + Thread.currentThread())

                fruitService.getFruitsSimulateOk()
            }

            awaitMain {
                when (val result = deferredResult.await()) {
                    is Fail -> handleFailure(failureWithPayload, result.value)
                    is Success -> handleSuccess(success, result.value)
                }
            }
        }

    }


    /**
     * identical to fetchFruitsAsync() but for demo purposes the URL we point to will give us an error,
     * we also don't specify a custom error class here
     */
    fun fetchFruitsButFail(
        success: SuccessCallback,
        failureWithPayload: FailureCallback<ErrorMessage>
    ) {

        logger.i("fetchFruitsButFail()")

        if (isBusy) {
            failureWithPayload(ErrorMessage.ERROR_BUSY)
            return
        }

        isBusy = true
        notifyObservers()


        launchMain {

            val result = callWrapper.processCallAwait {
                fruitService.getFruitsSimulateNotAuthorised()
            }

            awaitMain {
                when (result) {
                    is Fail -> handleFailure(failureWithPayload, result.value)
                    is Success -> handleSuccess(success, result.value)
                }
            }
        }
    }


    /**
     * identical to fetchFruitsAsync() but for demo purposes the URL we point to will give us an error,
     * here we specify a custom error class for more detail about the error than just an HTTP code can give us
     */
    fun fetchFruitsButFailAdvanced(
        success: SuccessCallback,
        failureWithPayload: FailureCallback<ErrorMessage>
    ) {

        logger.i("fetchFruitsButFailAdvanced()")

        if (isBusy) {
            failureWithPayload(ErrorMessage.ERROR_BUSY)
            return
        }

        isBusy = true
        notifyObservers()

        launchMain {

            val result = callWrapper.processCallAwait(FruitsCustomError::class.java) {
                fruitService.getFruitsSimulateNotAuthorised()
            }

            awaitMain {
                when (result) {
                    is Fail -> handleFailure(failureWithPayload, result.value)
                    is Success -> handleSuccess(success, result.value)
                }
            }
        }
    }


    /**
     * Demonstration of how to use carryOn to chain multiple connection requests together in a
     * simple way
     */
    fun chainedCall(
        success: SuccessCallback,
        failureWithPayload: FailureCallback<ErrorMessage>
    ) {

        logger.i("chainedCall()")

        if (isBusy) {
            failureWithPayload(ErrorMessage.ERROR_BUSY)
            return
        }

        isBusy = true
        notifyObservers()

        launchMain {

            val result = callWrapper.processCallAwait(
                FruitsCustomError::class.java
            ) {

                /**
                 * we're using fore's carryOn() extension function here but you can do whatever
                 * you like here, including using reactive streams if appropriate. Once your network
                 * chain is complete though, when you want to expose the resulting state,
                 * you need to: 1) set it, and 2) call notifyObservers().
                 *
                 * (carryOn() lets you transparently handle networking
                 * errors at each step - Internet search for "railway oriented programming"
                 * or "andThen" functions)
                 */
                var ticketRef = ""
                logger.i("...create user...")
                fruitService.createUser()
                    .carryOn {
                        logger.i("...create user ticket...")
                        fruitService.createUserTicket(it.userId)
                    }
                    .carryOn {
                        ticketRef = it.ticketRef
                        logger.i("...get waiting time...")
                        fruitService.getEstimatedWaitingTime(it.ticketRef)
                    }
                    .carryOn {
                        if (it.minutesWait > 10) {
                            logger.i("...cancel ticket...")
                            fruitService.cancelTicket(ticketRef)
                        } else {
                            logger.i("...confirm ticket...")
                            fruitService.confirmTicket(ticketRef)
                        }
                    }
                    .carryOn {
                        logger.i("...claim free fruit!...")
                        fruitService.claimFreeFruit(it.ticketRef)
                    }
            }

            awaitMain {
                when (result) {
                    is Fail -> handleFailure(failureWithPayload, result.value)
                    is Success -> handleSuccess(success, result.value)
                }
            }
        }
    }

    private fun handleSuccess(
        success: SuccessCallback,
        successResponse: List<FruitPojo>
    ) {

        logger.i("handleSuccess() t:" + Thread.currentThread())

        currentFruit = selectRandomFruit(successResponse)
        success()
        complete()
    }

    private fun handleFailure(
        failureWithPayload: FailureCallback<ErrorMessage>,
        failureMessage: ErrorMessage
    ) {

        logger.i("handleFailure() t:" + Thread.currentThread())

        failureWithPayload(failureMessage)
        complete()
    }

    private fun complete() {

        logger.i("complete() t:" + Thread.currentThread())

        isBusy = false
        notifyObservers()
    }

    private fun selectRandomFruit(listOfFruits: List<FruitPojo>): FruitPojo {
        return listOfFruits[if (listOfFruits.size == 1) 0 else random.nextInt(listOfFruits.size - 1)]
    }

    companion object {
        private val random = Random()
    }
}
